### ISSUE TITLE
[PATCH] TODO test for AUTOLOAD XSUB and inheritance

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10158]
Migrated from [rt.perl.org#72708](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72708) (status was 'resolved')

Searchable as RT72708$

**Comment by p5pRT** [comment]
### From @tsee

#### Created by tsee@l3tsee.nonet



If an AUTOLOAD sub is implemented in XS\, $AUTOLOAD will not be set\. This 
was done as an optimization\. Instead\, the PV of the XSUB's CV is set to 
the name of the method that was called\. But it's not possible for the 
AUTOLOAD XSUB to find out which \*class\* the method was called in 
originally\. This \-\- at the very least \-\- breaks inheritance\.

Attached is a patch that implements a TODO test for AUTOLOAD as an XSUB 
and inheritance\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl 5.11.4:

Configured by tsee at Wed Feb 10 23:45:17 CET 2010.

Summary of my perl5 (revision 5 version 11 subversion 4) configuration:
   Commit id: b9ad13acb338e137b9560a8b578e1f7c983706be
   Platform:
     osname=linux, osvers=2.6.31-19-generic, 
archname=x86_64-linux-thread-multi
     uname='linux l3tsee 2.6.31-19-generic #56-ubuntu smp thu jan 28 
02:39:34 utc 2010 x86_64 gnulinux '
     config_args='-Doptimize=-g -Dprefix=/home/tsee/bleadperl/install 
-Dusedevel -Dcc=ccache gcc -Dld=gcc -Dusethreads -DDEBUGGUNG -de'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=define, use64bitall=define, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='ccache gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING 
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include 
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
     optimize='-g',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING 
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
     ccversion='', gccversion='4.4.1', gccosandvers=''
     intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
     libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
     perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
     libc=/lib/libc-2.10.1.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.10.1'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
     cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib 
-fstack-protector'

Locally applied patches:



@INC for perl 5.11.4:
 
/home/tsee/bleadperl/install/lib/site_perl/5.11.4/x86_64-linux-thread-multi
     /home/tsee/bleadperl/install/lib/site_perl/5.11.4
     /home/tsee/bleadperl/install/lib/5.11.4/x86_64-linux-thread-multi
     /home/tsee/bleadperl/install/lib/5.11.4
     /home/tsee/bleadperl/install/lib/site_perl
     .


Environment for perl 5.11.4:
     HOME=/home/tsee
     LANG=de_DE.UTF-8
     LANGUAGE=
     LD_LIBRARY_PATH=/home/tsee/src/mylib/lib:
     LOGDIR (unset)
 
PATH=/home/tsee/bleadperl/install/bin/:/home/tsee/bin:/home/tsee/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @tsee

<details><summary>0001-TODO-test-for-inheritance-with-XS-AUTOLOAD.patch</summary>

```diff
From 8413fbf6febde5444b4f4188cf5dbd446aac955b Mon Sep 17 00:00:00 2001
From: Steffen Mueller <smueller@cpan.org>
Date: Thu, 11 Feb 2010 00:29:44 +0100
Subject: [PATCH] TODO test for inheritance with XS AUTOLOAD

If an AUTOLOAD sub is an XSUB, $AUTOLOAD won't be set. This is intended
as an optimization, but the replacement of setting the PV of the XSUB's
CV to the method name (without class) breaks inheritance because the
XSUB has no way of finding out how it was called. Abusing the PV of the
CV is a bad idea anyway (prototype!).
---
 ext/XS-APItest/APItest.xs     |   25 +++++++++++
 ext/XS-APItest/t/xsautoload.t |   91 +++++++++++++++++++++++++++++++++++++++++
 2 files changed, 116 insertions(+), 0 deletions(-)
 create mode 100644 ext/XS-APItest/t/xsautoload.t

diff --git a/ext/XS-APItest/APItest.xs b/ext/XS-APItest/APItest.xs
index ede6994..744b0d4 100644
--- a/ext/XS-APItest/APItest.xs
+++ b/ext/XS-APItest/APItest.xs
@@ -936,3 +936,28 @@ void
 my_exit(int exitcode)
         PPCODE:
         my_exit(exitcode);
+
+MODULE = XS::APItest PACKAGE = XS::APItest::AUTOLOADtest
+
+int
+AUTOLOAD(...)
+  INIT:
+    SV* comms;
+    STRLEN len;
+    SV* class_and_method;
+    SV* tmp;
+  CODE:
+    class_and_method = get_sv("AUTOLOAD", 0);
+    comms = get_sv("main::the_method", 1);
+    if (class_and_method == NULL) {
+      RETVAL = 1;
+    } else if (!SvOK(class_and_method)) {
+      RETVAL = 2;
+    } else if (!SvPOK(class_and_method)) {
+      RETVAL = 3;
+    } else {
+      SvPV_set(comms, SvPV(class_and_method, len));
+      RETVAL = 0;
+    }
+  OUTPUT: RETVAL
+
diff --git a/ext/XS-APItest/t/xsautoload.t b/ext/XS-APItest/t/xsautoload.t
new file mode 100644
index 0000000..0d32220
--- /dev/null
+++ b/ext/XS-APItest/t/xsautoload.t
@@ -0,0 +1,91 @@
+#!perl -w
+BEGIN {
+    push @INC, "::lib:$MacPerl::Architecture:" if $^O eq 'MacOS';
+    require Config; import Config;
+    if ($Config{'extensions'} !~ /\bXS\/APItest\b/) {
+        # Look, I'm using this fully-qualified variable more than once!
+        my $arch = $MacPerl::Architecture;
+        print "1..0 # Skip: XS::APItest was not built\n";
+        exit 0;
+    }
+    $| = 1;
+}
+
+#
+# This is a test for AUTOLOAD implemented as an XSUB.
+# It tests that $AUTOLOAD is set correctly, including the
+# case of inheritance.
+#
+# Rationale: Due to change adb5a9ae, $AUTOLOAD is not currently set
+# for XSUB AUTOLOADs at all. Instead, the PV of the AUTOLOAD
+# XSUB is set to the name of the method. This makes inheriting
+# an XSUB AUTOLOAD impossible since it's impossible to figure
+# out in which class it was called in the first place.
+# We cruelly test it regardless.
+# 
+
+use strict;
+use Test::More tests => '12';
+
+use_ok('XS::APItest');
+# First, make sure we have the XS AUTOLOAD available for testing
+ok(XS::APItest::AUTOLOADtest->can('AUTOLOAD'), 'Test class ->can AUTOLOAD');
+
+# Used to communicate from the XS AUTOLOAD to Perl land
+use vars '$the_method';
+
+# First, set up the Perl equivalent to what we're testing in
+# XS so we have a comparison
+package PerlBase;
+use vars '$AUTOLOAD';
+sub AUTOLOAD {
+  Test::More::ok(defined $AUTOLOAD);
+  return 1 if not defined $AUTOLOAD;
+  $main::the_method = $AUTOLOAD;
+  return 0;
+}
+
+package PerlDerived;
+use vars '@ISA';
+@ISA = qw(PerlBase);
+
+package Derived;
+use vars '@ISA';
+@ISA = qw(XS::APItest::AUTOLOADtest);
+
+package main;
+
+# Test Perl AUTOLOAD in base class directly
+$the_method = undef;
+is(PerlBase->Blah(), 0,
+   "Perl AUTOLOAD gets called and returns success");
+is($the_method, 'PerlBase::Blah',
+   'Scalar set to correct class/method name');
+
+# Test Perl AUTOLOAD in derived class
+$the_method = undef;
+is(PerlDerived->Boo(), 0,
+   'Perl AUTOLOAD on derived class gets called and returns success');
+is($the_method, 'PerlDerived::Boo',
+   'Scalar set to correct class/method name');
+
+# Test XS AUTOLOAD in base class directly
+$the_method = undef;
+TODO: {
+  local $TODO = 'Bug: $AUTOLOAD not set for XSUB AUTOLOADs';
+  is(XS::APItest::AUTOLOADtest->Blah(), 0,
+     'XS AUTOLOAD gets called and returns success');
+  is($the_method, 'XS::APItest::AUTOLOADtest::Blah',
+     'Scalar set to correct class/method name');
+}
+
+# Test XS AUTOLOAD in derived class directly
+$the_method = undef;
+TODO: {
+  local $TODO = 'Bug: $AUTOLOAD not set for XSUB AUTOLOADs';
+  is(Derived->Foo(), 0,
+     'XS AUTOLOAD gets called and returns success');
+  is($the_method, 'Derived::Foo',
+     'Scalar set to correct class/method name');
+}
+
-- 
1.6.3.3


```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout


On Oct 11\, 2011\, at 9&#8203;:10 AM\, Father Chrysostomos wrote&#8203;:

> 
> On Oct 11\, 2011\, at 9&#8203;:03 AM\, Steffen Mueller wrote&#8203;:
> 
>> This is regarding Father Chrysostomos' changes to fix the current behaviour of XS AUTOLOAD methods\, specifically commits 8fa6a40953ef88573ed3cbbb37666e7b72dec7dd and the tests in 6911735f50121ad015d280f86e257e8e9eae797a\.
>> 
>> My network connection is too slow/bad/mobile to searching for the bug\, but I know there's a perlbug that I filed about AUTOLOAD methods implemented in XS\.

https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72708

>> In a nutshell\, because Doug's hack prevents $AUTOLOAD from being set at all\, the XS AUTOLOAD cannot be inherited/reused\. The SvPVX of the CV only contains the method name\, not the class\, so it's not sufficient for figuring out for which class the AUTOLOAD was invoked\.
> 
> It’s not documented\, but you can use CvSTASH\.

In view of that use of CvSTASH\, is it necessary to set $AUTOLOAD?  Regardless of whether it is necessary\, is it desirable? :\-\)

>> 
>> I hope my memory didn't fail me\, but it'll be some time before I can check for sure\.
>> 
>> This bug in conjunction with the general badness of the SvPVX\(CV\) abuse makes me think that unless a CPAN grep turns up many examples\, it may be worth breaking the long\-standing behaviour\. If we don't\, does anybody have an idea how to efficiently include the namespace part of $AUTOLOAD?
>> 
>> Best regards\,
>> Steffen
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 11 09&#8203;:22&#8203;:59 2011\, sprout wrote&#8203;:
> 
> On Oct 11\, 2011\, at 9&#8203;:10 AM\, Father Chrysostomos wrote&#8203;:
> 
> >
> > On Oct 11\, 2011\, at 9&#8203;:03 AM\, Steffen Mueller wrote&#8203;:
> >
> >> This is regarding Father Chrysostomos' changes to fix the current
>    behaviour of XS AUTOLOAD methods\, specifically commits
>    8fa6a40953ef88573ed3cbbb37666e7b72dec7dd and the tests in
>    6911735f50121ad015d280f86e257e8e9eae797a\.
> >>
> >> My network connection is too slow/bad/mobile to searching for the
>    bug\, but I know there's a perlbug that I filed about AUTOLOAD
>    methods implemented in XS\.
> 
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72708
> 
> >> In a nutshell\, because Doug's hack prevents $AUTOLOAD from being
>    set at all\, the XS AUTOLOAD cannot be inherited/reused\. The SvPVX
>    of the CV only contains the method name\, not the class\, so it's not
>    sufficient for figuring out for which class the AUTOLOAD was
>    invoked\.
> >
> > It’s not documented\, but you can use CvSTASH\.
> 
> In view of that use of CvSTASH\, is it necessary to set $AUTOLOAD?
>    Regardless of whether it is necessary\, is it desirable? :\-\)

I’ve just discovered bug \#6828\, according to which $AUTOLOAD \*was\*
populated in 5\.6\.0\, but not in 5\.6\.1\.

A diff between the two shows this in gv\_autoload4&#8203;:

\+    if \(\!CvROOT\(cv\)\)
\+    return Nullgv;

Back then\, CvROOT and CvXSUB were two separate fields in the struct\, so
that meant XS autoloading stopped working\.

So it was reintroduced in 5\.8\.0 as a ‘new feature’\.

I find that a very compelling reason to reinstate the old behaviour\. 
But the new behaviour has been around long enough we probably have to
keep both\.

> 
> >>
> >> I hope my memory didn't fail me\, but it'll be some time before I
>    can check for sure\.
> >>
> >> This bug in conjunction with the general badness of the SvPVX\(CV\)
>    abuse makes me think that unless a CPAN grep turns up many
>    examples\, it may be worth breaking the long\-standing behaviour\. If
>    we don't\, does anybody have an idea how to efficiently include the
>    namespace part of $AUTOLOAD?
> >>
> >> Best regards\,
> >> Steffen
> >
>

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Oct 11\, 2011 at 09&#8203;:44&#8203;:41AM \-0700\, Father Chrysostomos via RT wrote&#8203;:

> A diff between the two shows this in gv\_autoload4&#8203;:
> 
> \+    if \(\!CvROOT\(cv\)\)
> \+    return Nullgv;
> 
> Back then\, CvROOT and CvXSUB were two separate fields in the struct\, so
> that meant XS autoloading stopped working\.
> 
> So it was reintroduced in 5\.8\.0 as a 'new feature'\.

I'm on a very poor Internet connection currently \[and supposed to be on
holiday :\-\)\] so I'm not really able to follow the detail of this this week
\(or next\, I suspect\)\, but \(from memory\) CvROOT\(\) and CvXSUB\(\) were merged
between 5\.8\.0 and 5\.10\.0\. Which suggests that that merger can't be
responsible for changes between 5\.60 and 5\.8\.0

> I find that a very compelling reason to reinstate the old behaviour\. 
> But the new behaviour has been around long enough we probably have to
> keep both\.

If I have it right\, \*both\* behaviours have been around for over a decade\.
This is a mess\.

Not sure if either has been documented until recently\, but the use of SvPVX
to store the prototype is \(I think\) more well known\. Almost no\-one writes
AUTOLOADers in XS\. Far more people introspect the symbol table\, look at
Devel&#8203;::Peek output \(and potentially the code behind it\)\, and want to
manipulate prototypes\.

The one get out I think might help here is that it makes no sense to have a
prototype on AUTOLOAD\. It's called as a method\. Hence it might be the least
worst solution to have a runtime check and one of

1&#8203;: forbid using an XS AUTOLOAD with a prototype
2&#8203;: only store one of $AUTOLOAD or the prototype in a "special" way if it's
&nbsp;   an XS subroutine with a prototype
3&#8203;: if there is a prototype\, use the save stack to save it whilst calling the
&nbsp;   subroutine\. \[Yes\, that make a speed penalty for the "no sane code would
&nbsp;   do this" case\, but\, well no sane code should do it\, \*and\* it won't break
&nbsp;   an the implementation of an existing XS AUTOLOAD if someone monkeys with
&nbsp;   it and adds a prototype\. Then again\, maybe it \*is\* fair game to break if
&nbsp;   someone things it fair and reasonable to monkey with the prototype of
&nbsp;   someone else's code\. In which case\, I'm liking option 1 the best currently\]

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 11 12&#8203;:23&#8203;:15 2011\, nicholas wrote&#8203;:
> On Tue\, Oct 11\, 2011 at 09&#8203;:44&#8203;:41AM \-0700\, Father Chrysostomos via RT
> wrote&#8203;:
> 
> > A diff between the two shows this in gv\_autoload4&#8203;:
> >
> > \+    if \(\!CvROOT\(cv\)\)
> > \+    return Nullgv;
> >
> > Back then\, CvROOT and CvXSUB were two separate fields in the struct\,
> so
> > that meant XS autoloading stopped working\.
> >
> > So it was reintroduced in 5\.8\.0 as a 'new feature'\.
> 
> I'm on a very poor Internet connection currently \[and supposed to be
> on
> holiday :\-\)\] so I'm not really able to follow the detail of this this
> week
> \(or next\, I suspect\)\, but \(from memory\) CvROOT\(\) and CvXSUB\(\) were
> merged
> between 5\.8\.0 and 5\.10\.0\. Which suggests that that merger can't be
> responsible for changes between 5\.60 and 5\.8\.0

No\, that was a diff between 5\.6\.0 and 5\.6\.1\.  The fact that they hadn’t
been merged yet was what caused that change to introduce a bug\.

Here’s a timeline for you&#8203;:

5\.6\.0 \- XS AUTOLOAD works with $AUTOLOAD
5\.6\.1 \- AUTOLOAD no longer tries to call an undefined AUTOLOAD stub;
&nbsp;        a side effect is that XS AUTOLOAD no longer works at all
5\.8\.0 \- XS AUTOLOAD is made to work again\, as a ‘new feature’\.  It
&nbsp;        uses SvPVX instead of $AUTOLOAD\, ‘so the XSUB won’t have to
&nbsp;        fetch and split it’\.
5\.15\.4\-to\-be \- I notice that it conflicts with prototypes\, so I fix
&nbsp;               it in the most backward\-compatible way I can\.

> 
> > I find that a very compelling reason to reinstate the old behaviour\.
> > But the new behaviour has been around long enough we probably have
> to
> > keep both\.
> 
> If I have it right\, \*both\* behaviours have been around for over a
> decade\.
> This is a mess\.
> 
> Not sure if either has been documented until recently\, but the use of
> SvPVX
> to store the prototype is \(I think\) more well known\.

But anyone relying on it too much will have run into bugs \(sneaky CvFILE
allocation\, XS AUTOLOAD\, etc\.\)\, so it can’t be used very much\.  It’s
still not documented anywhere\, so I think we are ‘allowed’ to break it\.

> Almost no\-one
> writes
> AUTOLOADers in XS\. Far more people introspect the symbol table\, look
> at
> Devel&#8203;::Peek output \(and potentially the code behind it\)\, and want to
> manipulate prototypes\.
> 
> The one get out I think might help here is that it makes no sense to
> have a
> prototype on AUTOLOAD\. It's called as a method\. Hence it might be the
> least
> worst solution to have a runtime check and one of
> 
> 1&#8203;: forbid using an XS AUTOLOAD with a prototype
> 2&#8203;: only store one of $AUTOLOAD or the prototype in a "special" way if
> it's
>    an XS subroutine with a prototype

Number 2 is what I have already done\.  It should be fully
backward\-compatible\, so I see no reason to break it\.

I even went as far as to special\-case sv\_setpv to keep Scalar&#8203;::Util and
Sub&#8203;::Prototype working\.

> 3&#8203;: if there is a prototype\, use the save stack to save it whilst
> calling the
>    subroutine\. \[Yes\, that make a speed penalty for the "no sane code
> would
>    do this" case\, but\, well no sane code should do it\, \*and\* it won't
> break
>    an the implementation of an existing XS AUTOLOAD if someone monkeys
> with
>    it and adds a prototype\. Then again\, maybe it \*is\* fair game to
> break if
>    someone things it fair and reasonable to monkey with the prototype
> of
>    someone else's code\. In which case\, I'm liking option 1 the best
> currently\.

Except option 1 requires more work and is likely to break code\.

Implementing both $AUTOLOAD and SvPVX requires changing about two lines
of code\.  I’ve almost finished\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Feb 10 15&#8203;:42&#8203;:29 2010\, smueller@&#8203;cpan\.org wrote&#8203;:
> This is a bug report for perl from tsee@&#8203;l3tsee\.nonet\,
> generated with the help of perlbug 1\.39 running under perl 5\.11\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> If an AUTOLOAD sub is implemented in XS\, $AUTOLOAD will not be set\.
> This
> was done as an optimization\. Instead\, the PV of the XSUB's CV is set
> to
> the name of the method that was called\. But it's not possible for the
> AUTOLOAD XSUB to find out which \*class\* the method was called in
> originally\. This \-\- at the very least \-\- breaks inheritance\.
> 
> Attached is a patch that implements a TODO test for AUTOLOAD as an
> XSUB
> and inheritance\.

The comment in that patch says that commit adb5a9ae stopped $AUTOLOAD
from being set\, but it was actually ed850460\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 11 14&#8203;:53&#8203;:41 2011\, sprout wrote&#8203;:
> On Tue Oct 11 12&#8203;:23&#8203;:15 2011\, nicholas wrote&#8203;:
> > 1&#8203;: forbid using an XS AUTOLOAD with a prototype

> Except option 1 requires more work and is likely to break code\.
> 
> Implementing both $AUTOLOAD and SvPVX requires changing about two lines
> of code\.  I’ve almost finished\.

I’m actually more wary of changing documented behaviour than breaking
code\.  I haven’t found an instance of XS AUTOLOAD on CPAN yet\, but I
lost my connection to dromedary where I’m searching and had to start
over\. :\-\(

But the main thing is this&#8203;:  Changing undocumented behaviour is one
thing; if it breaks a few modules that’s probably OK\.  But changing
\*documented\* behaviour just doesn’t seem right\, at least not without a
deprecation cycle\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Feb 10 15&#8203;:42&#8203;:29 2010\, smueller@&#8203;cpan\.org wrote&#8203;:
> This is a bug report for perl from tsee@&#8203;l3tsee\.nonet\,
> generated with the help of perlbug 1\.39 running under perl 5\.11\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> If an AUTOLOAD sub is implemented in XS\, $AUTOLOAD will not be set\.
> This
> was done as an optimization\. Instead\, the PV of the XSUB's CV is set
> to
> the name of the method that was called\. But it's not possible for the
> AUTOLOAD XSUB to find out which \*class\* the method was called in
> originally\. This \-\- at the very least \-\- breaks inheritance\.
> 
> Attached is a patch that implements a TODO test for AUTOLOAD as an
> XSUB
> and inheritance\.

I’ve just munged it and applied it as 120b7a0\.  Fix coming in a few
hours\.\.\.\. \(I’ve just run out of time for the time being\.\)

I’m marking this ticket as resolved\, as we still have \#6828 open for
$AUTOLOAD’s not being set\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Oct 11 18&#8203;:22&#8203;:36 2011\, sprout wrote&#8203;:
> On Wed Feb 10 15&#8203;:42&#8203;:29 2010\, smueller@&#8203;cpan\.org wrote&#8203;:
> > This is a bug report for perl from tsee@&#8203;l3tsee\.nonet\,
> > generated with the help of perlbug 1\.39 running under perl 5\.11\.4\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please describe your issue here\]
> > 
> > If an AUTOLOAD sub is implemented in XS\, $AUTOLOAD will not be set\.
> > This
> > was done as an optimization\. Instead\, the PV of the XSUB's CV is set
> > to
> > the name of the method that was called\. But it's not possible for the
> > AUTOLOAD XSUB to find out which \*class\* the method was called in
> > originally\. This \-\- at the very least \-\- breaks inheritance\.
> > 
> > Attached is a patch that implements a TODO test for AUTOLOAD as an
> > XSUB
> > and inheritance\.
> 
> I’ve just munged it and applied it as 120b7a0\.  Fix coming in a few
> hours\.\.\.\. \(I’ve just run out of time for the time being\.\)
> 
> I’m marking this ticket as resolved\, as we still have \#6828 open for
> $AUTOLOAD’s not being set\.

I’ve just fixed that with commit bb619f3\.

On a related note\, would you be willing to smoke CPAN with commit
6911735f \(before my XS AUTOLOAD/prototype changes\) and 74ee33f2 \(the
last change\)?

**Comment by p5pRT** [comment]
### From @tsee

On 10/11/2011 06&#8203;:22 PM\, Father Chrysostomos wrote&#8203;:
> In view of that use of CvSTASH\, is it necessary to set $AUTOLOAD?
> Regardless of whether it is necessary\, is it desirable? :\-\)

I guess not\. I missed that bit\.

Thanks for the advice\!

Cheers\,
Steffen

**Comment by p5pRT** [comment]
### From @tsee

On 10/12/2011 06&#8203;:40 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Tue Oct 11 18&#8203;:22&#8203;:36 2011\, sprout wrote&#8203;:
>> On Wed Feb 10 15&#8203;:42&#8203;:29 2010\, smueller@&#8203;cpan\.org wrote&#8203;:
>>> This is a bug report for perl from tsee@&#8203;l3tsee\.nonet\,
>>> generated with the help of perlbug 1\.39 running under perl 5\.11\.4\.
>>>
>>>
>>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>> \[Please describe your issue here\]
>>>
>>> If an AUTOLOAD sub is implemented in XS\, $AUTOLOAD will not be set\.
>>> This
>>> was done as an optimization\. Instead\, the PV of the XSUB's CV is set
>>> to
>>> the name of the method that was called\. But it's not possible for the
>>> AUTOLOAD XSUB to find out which \*class\* the method was called in
>>> originally\. This \-\- at the very least \-\- breaks inheritance\.
>>>
>>> Attached is a patch that implements a TODO test for AUTOLOAD as an
>>> XSUB
>>> and inheritance\.
>>
>> I’ve just munged it and applied it as 120b7a0\.  Fix coming in a few
>> hours\.\.\.\. \(I’ve just run out of time for the time being\.\)
>>
>> I’m marking this ticket as resolved\, as we still have \#6828 open for
>> $AUTOLOAD’s not being set\.
>
> I’ve just fixed that with commit bb619f3\.
>
> On a related note\, would you be willing to smoke CPAN with commit
> 6911735f \(before my XS AUTOLOAD/prototype changes\) and 74ee33f2 \(the
> last change\)?

Of course\. I'll try to kick it off asap\.

Best regards\,
Steffen

**Comment by p5pRT** [comment]
### From @tsee

On 10/12/2011 06&#8203;:40 AM\, Father Chrysostomos via RT wrote&#8203;:
> On a related note\, would you be willing to smoke CPAN with commit
> 6911735f \(before my XS AUTOLOAD/prototype changes\) and 74ee33f2 \(the
> last change\)?

The smoke machine is building the two perls with the required modules 
now\. I'll kick off the actual smokers when I get in the office\.

\-\-Steffen

PS&#8203;: I didn't update the minicpan\, so that's moderately outdated \(4 
weeks?\)\. Presumably that's not a big deal at all and I may fix that 
quickly from the office\.
